src/components/core/NotifyObservers.react.js

```diff
@@ -17,79 +17,49 @@ function mapStateToProps(state) {
+            // Identify the modified props that are required for callbacks
+            const watchedKeys = filter(key =>
+                stateProps.dependencies &&
+                stateProps.dependencies.find(dependency =>
+                    dependency.inputs.find(input => input.id === ownProps.id && input.property === key) ||
+                    dependency.state.find(state => state.id === ownProps.id && state.property === key)
+                )
+            )(keysIn(newProps));
+
+            // Always update this component's props
+            dispatch(updateProps({
                 props: newProps,
                 id: ownProps.id,
-                itempath: stateProps.paths[ownProps.id],
-            };
+                itempath: stateProps.paths[ownProps.id]
+            }));
 
-            // Update this component's props
-            dispatch(updateProps(payload));
+            // Only dispatch changes to Dash if a watched prop changed
+            if (watchedKeys.length) {
+                dispatch(notifyObservers({
+                    id: ownProps.id,
+                    props: pick(watchedKeys)(newProps)
+                }));
+            }
 
-            // Update output components that depend on this input
-            dispatch(notifyObservers({id: ownProps.id, props: newProps}));
         },
     };
 }
 
-function NotifyObserversComponent({
-    children,
-    id,
-    paths,
-
-    dependencies,
-
-    setProps,
-}) {
-    const thisComponentSharesState =
-        dependencies &&
-        dependencies.find(
-            dependency =>
-                dependency.inputs.find(input => input.id === id) ||
-                dependency.state.find(state => state.id === id)
-        );
-    /*
-     * Only pass in `setProps` if necessary.
-     * This allows component authors to skip computing unneeded data
-     * for `setProps`, which can be expensive.
-     * For example, consider `hoverData` for graphs. If it isn't
-     * actually used, then the component author can skip binding
-     * the events for the component.
-     *
-     * TODO - A nice enhancement would be to pass in the actual
-     * properties that are used into the component so that the
-     * component author can check for something like
-     * `subscribed_properties` instead of just `setProps`.
-     */
-    const extraProps = {};
-    if (
-        thisComponentSharesState &&
-        // there is a bug with graphs right now where
-        // the restyle listener gets assigned with a
-        // setProps function that was created before
-        // the item was added. only pass in setProps
-        // if the item's path exists for now.
-        paths[id]
-    ) {
-        extraProps.setProps = setProps;
-    }
-
-    if (!isEmpty(extraProps)) {
-        return React.cloneElement(children, extraProps);
-    }
-    return children;
+function NotifyObserversComponent({ children, setProps }) {
+    return React.cloneElement(children, { setProps });
```

Comment:
Always pass the setProps function to the component. It will censure itself as needed.